src/core/model.jl

```diff
@@ -107,7 +107,11 @@ export random_agent, nagents, allagents, allids, nextid, seed!
 
 Return an agent given its ID.
 """
-Base.getindex(m::ABM, id::Integer) = m.agents[id]
+function Base.getindex(model::ABM, id::Integer) 
+    ind = findfirst(x -> x == id, model.agents.id)
```

Comment:
This is super slow, which is most likely causing your performance issues.

I just pushed up my old work on this [here](https://github.com/JuliaDynamics/Agents.jl/commit/9750323f0c2e030b888e0e31671c7157bc23dd04), which uses a `Dict{Int,Int}` index property. Take a look and see if it's of any assistance.

Reply:
Yeees, I know. :( But I couldn't think of a better way to find the index of the given id. I tried some other methods but findfirst was already the fastest implementation I could think of. The O(1) lookup of dicts is basically unbeatable. Sticking with a `Dict{Int,Int}` would require us to recreate the dict from scratch every time an agent gets added/deleted. Will try it out though.

Reply:
> require us to recreate the dict from scratch every time an agent gets added/deleted. Will try it out though.

Is that true? 

add_agent: would use a specific id, thus `m.index[id] = id`,
kill_agent: `delete_at!(m.index[id])`

I may have missed something there though.

Reply:
In a very simple model or right at initialisation, the statement `m.index[id] = id` would work. But we we want to map `id => index` in the dict and we know that `id == index` is not always true so we need to take care of that, I suppose? That's why I think recreating the dict is necessary after every addition/removal of an agent.

Thinking about it a bit more, once you delete an agent from the StructArray, all the rows after it will be altered as well (go up one row, i.e. have their index in the array reduced by 1). Let's assume that the rows of the StructArray are always sorted by `id` (which would be the default order without interference by users). Since we only ever delete one agent at a time, we could possibly do `m.index[id] -= 1` for the agents with higher ids. But adding is not so easy anymore afterwards because then `id == index` isn't necessarily the case.

Will need to think about this a bit more and try a few things out. (This screams "write some tests" at me...)

Reply:
Ah, yes, that's right&mdash;I remember having that though and thinking it pretty untenable. Although I'm not sure if such an operation is actually that bad knowing the alternative. We swap O(1) add/deletes for O(n), but searches stay O(1) and we get the speed up of StructArrays as the tradeoff.

How that fares overall for all our cases will remain to be seen.